sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/conditionalExpressions.scala

```diff
@@ -91,7 +91,7 @@ trait CaseWhenLike extends Expression {
 
   // both then and else expressions should be considered.
   def valueTypes: Seq[DataType] = (thenList ++ elseValue).map(_.dataType)
-  def valueTypesEqual: Boolean = valueTypes.distinct.size == 1
+  def valueTypesEqual: Boolean = valueTypes.map(_.asNullable).distinct.size == 1
```

Comment:
how about:
```
valueTypes.sliding(2, 1).forall {
  case (dt1, dt2) => DataType.equalsIgnoreNullability(dt1, dt2)
}
```

`DataType.equalsIgnoreNullability` is transitive so this should work.